crates/bevy_ecs/src/schedule/executor_parallel.rs

```diff
@@ -166,70 +166,124 @@ impl ParallelExecutor {
         systems: &'scope mut [ParallelSystemContainer],
         world: &'scope World,
     ) {
+        // These are used as a part of a unit test.
+        #[cfg(test)]
+        let mut started_systems = 0;
         #[cfg(feature = "trace")]
         let _span = bevy_utils::tracing::info_span!("prepare_systems").entered();
         self.should_run.clear();
         for (index, (system_data, system)) in
             self.system_metadata.iter_mut().zip(systems).enumerate()
         {
+            let should_run = system.should_run();
+            let can_start = should_run
+                && system_data.dependencies_total == 0
+                && Self::can_start_now(
+                    self.non_send_running,
+                    system_data,
+                    &self.active_archetype_component_access,
+                );
             // Spawn the system task.
-            if system.should_run() {
-                self.should_run.set(index, true);
-                let start_receiver = system_data.start_receiver.clone();
+            if should_run {
+                self.should_run.insert(index);
                 let finish_sender = self.finish_sender.clone();
                 let system = system.system_mut();
                 #[cfg(feature = "trace")] // NB: outside the task to get the TLS current span
                 let system_span = bevy_utils::tracing::info_span!("system", name = &*system.name());
                 #[cfg(feature = "trace")]
                 let overhead_span =
                     bevy_utils::tracing::info_span!("system overhead", name = &*system.name());
-                let task = async move {
-                    start_receiver
-                        .recv()
-                        .await
-                        .unwrap_or_else(|error| unreachable!("{}", error));
+
+                let mut run = move || {
                     #[cfg(feature = "trace")]
-                    let system_guard = system_span.enter();
+                    let _system_guard = system_span.enter();
                     unsafe { system.run_unsafe((), world) };
-                    #[cfg(feature = "trace")]
-                    drop(system_guard);
-                    finish_sender
-                        .send(index)
-                        .await
-                        .unwrap_or_else(|error| unreachable!("{}", error));
                 };
 
-                #[cfg(feature = "trace")]
-                let task = task.instrument(overhead_span);
-                if system_data.is_send {
-                    scope.spawn(task);
+                if can_start {
+                    let task = async move {
+                        run();
+                        finish_sender
+                            .send(index)
+                            .await
+                            .unwrap_or_else(|error| unreachable!("{}", error));
```

Comment:
This needs a comment explaining why this can never panic.

Reply:
Added a comment.